### ISSUE TITLE
All kinds of convenience methods for creating bindings in FluentBindingRule should have an overload that accepts a factory that can process a ValueBindingContext

### ISSUE THREAD

**Issue by macrogreg** [Azure/azure-webjobs-sdk#2488]
The `FluentBindingRule` class contains a number of `BindToXxx` methods that represent convenience APIs for creating attribute bindings for function parameters and for return values.

Two of such methods (`BindToInput(..)` and `BindToStream(..)`) have overloads that allow the value factory to inspect the `ValueBindingContext` while creating the value bound by an attribute. This enables a number of scenarios: For example, it provides access to the name of the function being invoked, and to the id of the particular invocation instance. These values are otherwise not accessible.

Unfortunately, other `BindToXxx` methods do not such overloads (`BindToCollector(..)` and `BindToValueProvider(..)`). 
This proposal is to add corresponding overloads.

We currently require the respective `BindToCollector(..)` overload for scenarios in the [Availability Monitoring Extension](https://github.com/Azure/azure-functions-availability-monitoring-extension). The `BindToValueProvider(..)` overload should be added for consistency. 

A candidate PR is readily available (https://github.com/Azure/azure-webjobs-sdk/pull/2489).

**Comment by macrogreg** [comment]
Heads-up @mathewc : This is the change we discussed previously.
Also heads-up @brettsam FYI.

**Comment by zmarty** [comment]
@macrogreg @mathewc  Any idea when this can be checked in and shipped? We would like to use ValueBindingContext inside BindToCollector for an internal MSFT project.

**Comment by zmarty** [comment]
@fabiocav We still need this in an internal Microsoft project :) Thanks!

**Comment by zmarty** [comment]
@fabiocav  We still need this :)

**Comment by mathewc** [comment]
@zmarty Sorry this got lost in the shuffle. Assuming you guys still need these overloads so you can remove the temp reflection hack I put together for Greg?

**Comment by zmarty** [comment]
Hi @mathewc. Yes, we definitely still need this. Also, out of curiosity can you point me to this: "temp reflection hack I put together for Greg"? You can send me an internal mail if you wish, just look for my name.

**Comment by v-anvari** [comment]
Hi @zmarty , Can you confirm if this still requires a triage/ follow up or can this issue be closed.

**Comment by zmarty** [comment]
> Hi @zmarty , Can you confirm if this still requires a triage/ follow up or can this issue be closed.

Yes, we need this. No need to keep asking, the answer is **Yes**.
